xmtpv3/crates/libxmtp-core/src/vmac.rs

```diff
@@ -0,0 +1,90 @@
+use xmtp_proto::xmtp::v3::message_contents::vmac_unsigned_public_key::{
+    Union, VodozemacCurve25519,
+};
+use xmtp_proto::xmtp::v3::message_contents::{
+    VmacAccountLinkedKey, VmacContactBundle, VmacDeviceLinkedKey, VmacUnsignedPublicKey,
+};
+
+use serde_json::json;
+use vodozemac::{
+    olm::{Account, SessionConfig},
+    Curve25519PublicKey,
+};
+
+// Can't impl From<Curve25519PublicKey> for VmacUnsignedPublicKey because of orphan rules
+// - Need to come back and write a better translation trait, we can discuss
+fn vmac_unsigned_public_key_from_curve25519_public_key(
+    key: Curve25519PublicKey,
+) -> VmacUnsignedPublicKey {
+    VmacUnsignedPublicKey {
+        created_ns: 0,
+        union: Some(Union::Curve25519(VodozemacCurve25519 {
+            bytes: key.to_bytes().to_vec(),
+        })),
+    }
+}
+
+fn account_linked_key_from_curve25519_public_key(key: Curve25519PublicKey) -> VmacAccountLinkedKey {
```

Comment:
Wouldn't this be more idiomatic as a `impl from<Curve25519PublicKey> for VmacAccountLinkedKey {`

Reply:
I might be misinterpreting, but Rust's orphan rules don't allow me to write an Impl where both the trait and type live in different crates. Now that I think about it we maybe can work around it with either a Wrapper or a fake Generic type?

https://stackoverflow.com/q/75765502

Reply:
This is what the compiler spits at me:

```
error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> crates/libxmtp-core/src/vmac.rs:27:1
   |
27 | impl From<Curve25519PublicKey> for VmacUnsignedPublicKey {
   | ^^^^^-------------------------^^^^^---------------------
   | |    |                             |
   | |    |                             `VmacUnsignedPublicKey` is not defined in the current crate
   | |    `Curve25519PublicKey` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead
```

Reply:
I also totally skimmed past your comment above the first function here

Reply:
Very understandable it's a tiny comment punting a large problem haha

Reply:
Ya. A type wrapper would work, but I'm not sure of the downstream consequences of actually using the wrapped type.